### ISSUE TITLE
Timeout / Delay when getting token

### ISSUE THREAD

**Issue by Seb-Solon** [SUSE/Portus#653]
Hi there!

I have a docker based portus setup (registry etc are in a docker). I managed to get everything (at least what I need) working with ssl/https, pretty happy so far.

I am now experience some random latency when getting auth token .

Here are logs from the api : 

```
[13] IPSRC - - [06/Jan/2016:19:09:47 +0000] "GET /v2/token?account=BUILDACCOUNT&scope=repository%3ABUILDREP%2FPRODUCT%3Apush%2Cpull&service=REGISTRYSERVER%3A5000 HTTP/1.1" 200 
- 21.2648
```

On the other side (a automatic process pushing image to the registry) : 

```
Get http://REGISTRYSERVER:3000/v2/token?account=BUILDACCOUNT&scope=repository%3ABUILDREP%2FPRODUCT%3Apush%2Cpull&service=REGISTRYSERVER%3A5000: read tcp IPDEST:3000: use of closed network connection
```

Looks like the fact the get token takes +20s is an issue. Going a bit deeper, I found out other logs with 100s latency. 

I'm wondering what can get this token so long to generate. As far as I understand, the HTTP get is receive by the webapp and then some check are done (db request I imagine) and the token is generated right? Somewhere in this file I imagine : https://github.com/SUSE/Portus/blob/master/app/controllers/api/v2/tokens_controller.rb

Regards

**Comment by Seb-Solon** [comment]
Hi,

I did some investigations. 

First of all, the "use of closed network connection" is an error raised by go itself and was discussed in the following issue : https://github.com/golang/go/issues/9405 (the guy show that a slow server make the client raise this error). As mentioned in the ticket the fix for that is in Go > 1.5 , so very recent Go version. 

What I'm pointing here is that a solution may be get a docker version compiled with this version of Go, but it would still be a simple workaround what I'm facing (slow server)

Regarding Portus now, here are my findings : 

```
mkdir /tmp/testdocker/
cd /tmp/testdocker/

cat  Dockerfile
FROM busybox
RUN echo KK

for i in $(seq 20); do docker build --no-cache=true -t test1 . && docker tag test1 REGISTRYSERVER:5000/test/test1:$i && docker push REGISTRYSERVER:5000/test/test1:$i; done

```

I did the following with several version of docker : 

```
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 12:59:02 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 12:59:02 UTC 2015
 OS/Arch:      linux/amd64

AND

Client:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.5.1
 Git commit:   f4bf5c7
 Built:       
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.5.1
 Git commit:   f4bf5c7
 Built:       
 OS/Arch:      linux/amd64
```

The error **only** occurs on the first one, the one with the go version < 1.5 (see above)

Then, I removed the Portus "layer" by runnning another registry on the same server (port 5003) with only basic authentication (adapted from https://docs.docker.com/registry/deploying/) 

I **was not** able to reproduce the error in this case. So this make me think the slow part in somewhere in the Portus app. 


My setup is on a VM and the rails environment is development, maybe this causes to log more stuff and make the whole thing a bit slow?

I'm not very comfortable with Ruby so help would save me time. I would like to time so call function and understand where is the bottleneck. In my opinion Puma is working well, maybe the app is getting slow for wome reason, or maybe the MariaDB?

Anyway, I'm keep digging, will post new things.

**Comment by Seb-Solon** [comment]
Hi, there

Just you let you know that I fixed my issue by switching to a non "development" environment in Portus. By exporting RACK_ENV variable in the docker-compose to staging, configuration for staging is load. 

As suggested in the configuration file : 

```
  # In the development environment your application's code is reloaded on
  # every request. This slows down response time but is perfect for development
  # since you don't have to restart the web server when you make code changes.
```
